keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,997 @@
+// solution for such code.
+func Background() Logger {
+    return fallbackLogger
+}
+```
+
+#### logging helper API
+
+The following helper code will get copied from klog:
+- `ObjectRef`
+- `KMetaData`
+- `KObj`
+- `KRef`
+- `KObjs`
+
+To ensure that a single import is enough in source files, that package will also contain:
+
+```go
+type Logger = logr.Logger
+var NewContext = logr.NewContext
+// plus more as needed...
+```
+
+The implementation of `KObj` and `KObjs` will be either a direct copy of the
+code in klog or an improved version [with performance
+enhancements](https://github.com/kubernetes/kubernetes/issues/106945),
+depending [on benchmark](https://github.com/kubernetes/kubernetes/pull/106594)
+results.
+
+#### Logging in tests
+
+Ginkgo tests do not need to be changed. Because each process only runs one test
+at a time, the global default can be set and local Logger instances can be
+passed around just as in a normal binary.
+
+Unit tests with `go test` require a bit more work. Each test case must
+initialize a `logr/testing` Logger for its own instance of `testing.T`. The
+default log level will be 5, the level
+[recommended](https://github.com/kubernetes/community/blob/9406b4352fe2d5810cb21cc3cb059ce5886de157/contributors/devel/sig-instrumentation/logging.md#logging-conventions)
+for "the steps leading up to errors and warnings" and "for troubleshooting". It
+can be higher than in production binaries because `go test` without `-v` will
+not print the output for test cases that succeeded. It will only be shown for
+failed test cases, and for those the additional log messages may be useful to
+understand the failure. Optionally, logging options can be added to the test
+binary to modify this default log level.
+
+
+API:
+
+```go
+// NewTestContext returns a logger and context for use in a unit test case with
+// the default logging configuration. TestMainWithLogging can be used to enable
+// command line flags that modify those defaults.
+func NewTestContext(t *testing.T) (context.Context, Logger)
+
+// TestMainWithLogging can be used as a test's TestMain implementation.
+// It adds, parses and validates logging command line options before
+// running tests.
+func TestMainWithLogging(m *testing.M)
+```
+
+Example with additional command line options:
+
+```go
+import (
+    “testing”
+    ktesting “k8s.io/klogr/testing”
+)
+
+func TestMain(m *testing.M) {
+       ktesting.TestMainWithLogging(m)
+}
+
+func TestSomething(t *testing.T) {
+    // Either return value can be ignored with _ if not needed.
+    logger, ctx := ktesting.NewTestContext(t)
+            logger.Info(“test starts”)
+            doSomething(ctx)
+}
+```
+
+Custom log helper code must use the `WithCallStackHelper` method to ensure that
+the helper gets skipped during stack unwinding:
+
+```go
+func logSomething(logger log.Logger, obj interface{}) {
+    helper, logger := logger.WithCallStackHelper()
+    helper()
+    logger.Info(“I’m just helping”, “obj”, obj)
```

Comment:
You have angle quotes here which is breaking highlighting :)

Reply:
Wow, how did those get here?! I'm not even sure how to type them on my keyboard. Computers, always full of surprises. Fixed.